fluent-fallback/src/lib.rs

```diff
@@ -1,72 +1,257 @@
-use std::borrow::Borrow;
-use std::borrow::Cow;
+use cache::{AsyncCache, Cache};
+use fluent_bundle::{FluentArgs, FluentBundle, FluentError, FluentResource};
+use futures::Stream;
 
-use fluent_bundle::FluentResource;
-use fluent_bundle::{FluentArgs, FluentBundle};
+use std::{
+    borrow::{Borrow, Cow},
+    ops::{Deref, DerefMut},
+};
 
-use reiterate::Reiterate;
+mod cache;
 
-struct FluentBundleIterator<R, I>
+pub trait BundleGeneratorSync {
+    type Resource;
+    type Iter: Iterator<Item = FluentBundle<Self::Resource>>;
+
+    fn bundles_sync(&self, resource_ids: Vec<String>) -> Self::Iter;
+}
+
+pub trait BundleGenerator {
+    type Resource;
+    type Stream: Stream<Item = FluentBundle<Self::Resource>>;
+
+    fn bundles(&self, resource_ids: Vec<String>) -> Self::Stream;
+}
+
+pub struct L10nKey<'l> {
+    pub id: String,
+    pub args: Option<FluentArgs<'l>>,
+}
+
+pub struct SyncLocalization<G: BundleGeneratorSync> {
+    resource_ids: Vec<String>,
+    bundles: Cache<G::Iter>,
+    generator: G,
+}
+
+impl<G> SyncLocalization<G>
 where
-    I: Iterator<Item = FluentBundle<R>>,
+    G: BundleGeneratorSync + Default,
 {
-    iter: I,
+    pub fn new(resource_ids: Vec<String>) -> Self {
+        Self::with_generator(resource_ids, G::default())
+    }
 }
 
-impl<R, I> Iterator for FluentBundleIterator<R, I>
+impl<G> SyncLocalization<G>
 where
-    I: Iterator<Item = FluentBundle<R>>,
+    G: BundleGeneratorSync,
 {
-    type Item = Box<FluentBundle<R>>;
-    fn next(&mut self) -> Option<Self::Item> {
-        self.iter.next().map(Box::new)
+    pub fn with_generator(resource_ids: Vec<String>, generator: G) -> Self {
+        let bundles = Cache::new(generator.bundles_sync(resource_ids.clone()));
+
+        Self {
+            resource_ids,
+            bundles,
+            generator,
+        }
+    }
+
+    pub fn on_change(&mut self) {
+        // This invalidates the cache by recreating it.
+        self.bundles = Cache::new(self.generator.bundles_sync(self.resource_ids.clone()));
+    }
+
+    fn _format_value_sync_opt<'l>(
```

Comment:
I don't think you need to name it with `_` - rust knows its private because it doesn't have `pub`.

Reply:
Fixed.